hedera-node/src/main/java/com/hedera/services/store/contracts/SizeLimitedStorage.java

```diff
@@ -0,0 +1,331 @@
+import static com.hedera.services.exceptions.ValidationUtils.validateTrue;
+import static com.hedera.services.ledger.properties.AccountProperty.NUM_CONTRACT_KV_PAIRS;
+import static com.hedera.services.utils.EntityNum.fromLong;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.MAX_CONTRACT_STORAGE_EXCEEDED;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.MAX_STORAGE_IN_PRICE_REGIME_HAS_BEEN_USED;
+import static java.util.Objects.requireNonNull;
+import static org.apache.tuweni.units.bigints.UInt256.ZERO;
+
+/**
+ * Buffers a set of changes to the key/value pairs in contract storage into a <i>session</i>, and validates that their net
+ * effect will not cause any individual contract to exceed the limit set by {@link GlobalDynamicProperties#maxIndividualContractKvPairs()};
+ * nor the aggregate storage to exceed {@link GlobalDynamicProperties#maxAggregateContractKvPairs()}.
+ *
+ * Note that writing {@link UInt256#ZERO} to a key removes it from the map; so it is possible for a change to decrease
+ * the number of key/value pairs used.
+ */
+@Singleton
+public class SizeLimitedStorage {
+	public static final ContractValue ZERO_VALUE = ContractValue.from(ZERO);
+
+	/* Used to get the key/value storage limits */
+	private final GlobalDynamicProperties dynamicProperties;
+	/* Used to look up the initial key/value counts for the contracts involved in a change set */
+	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts;
+	/* Used to both read and write key/value pairs throughout the lifecycle of a change set */
+	private final Supplier<VirtualMap<ContractKey, ContractValue>> storage;
+
+	private final Map<Long, AtomicInteger> newUsages = new TreeMap<>();
+	private final Map<Long, TreeSet<ContractKey>> updatedKeys = new TreeMap<>();
+	private final Map<Long, TreeSet<ContractKey>> removedKeys = new TreeMap<>();
+	private final Map<ContractKey, ContractValue> newMappings = new HashMap<>();
+
+	private long totalKvPairs;
+
+	@Inject
+	public SizeLimitedStorage(
+			final GlobalDynamicProperties dynamicProperties,
+			final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts,
+			final Supplier<VirtualMap<ContractKey, ContractValue>> storage
+	) {
+		this.dynamicProperties = dynamicProperties;
+		this.accounts = accounts;
+		this.storage = storage;
+	}
+
+	/**
+	 * Clears all buffers and prepares for a new change-set of key/value pairs.
+	 */
+	public void beginSession() {
+		newUsages.clear();
+		updatedKeys.clear();
+		removedKeys.clear();
+		newMappings.clear();
+		/* We will update this count as changes are buffered throughout the session. */
+		totalKvPairs = storage.get().size();
+	}
+
+	/**
+	 * Validates that the pending key/value changes will not exceed any storage limits, and then
+	 * commits them to the underlying data source.
+	 *
+	 * @throws com.hedera.services.exceptions.InvalidTransactionException if a storage limit is exceeded
```

Comment:
This method doesn't throw, so the `@throws ` annotation can be removed ?

Reply:
Throwing an unchecked `InvalidTransactionException` is part of the contract for this method, so it's recommended to include in the Javadoc I believe.

Reply:
Ah I see. got it ! Thanks @tinker-michaelj